### PR TITLE
My Jetpack: Add "Expired" & "Expires soon" statuses to product cards

### LINKED ISSUE #39762: Backup Card: My Jetpack shows "Needs Plan" for backup even when there is a Jetpack plan from a hosting partner.
We've had a case where a user with an active Jetpack plan provided by hosting partners (e.g., including backup services) sees the "VaultPress Backup" section in the My Jetpack displaying "Needs plan." This causes confusion, leading users to believe they do not have an active backup, resulting in some purchasing an additional Jetpack Backup plan unnecessarily. 

Image

They had to contact us, cancel the plan from us and issue a refund. 

Is there a way we could improve the visibility of hosting-provided backup plans in the My Jetpack to avoid this kind of confusion?

This was reported in 8871069-zd-a8c

### PR DESCRIPTION
**Pull request by elliottprogrammer**
This PR adds an "Expired" or "Expiring soon" status to the product cards of products with an expired or expiring paid plan subscription. We also make the product CTA say "Renew my plan" (or "Resume my plan" if expired) and link to the specific subscription in purchase management section of their wordpress.com account so that they can renew or resume their subscription.

### Screenshots:

Screen Shot 2024-11-26 at 08 12 58

Screen Shot 2024-10-22 at 18 57 57

.

## Proposed changes:

* Added an "Expired" or "Expiring soon" status to the My Jetpack product cards.
* For products that are Expired, the CTA says "Resume my plan" and points to the wordpress.com purchase management page. For products that are Expiring soon, the CTA says "Renew my plan", and points directly to checkout with the product renewal in the cart.
* Also fixes issue: [Backup Card: My Jetpack shows "Needs Plan" for backup even when there is a Jetpack plan from a hosting partner](https://github.com/Automattic/jetpack/issues/39762). - Testing instructions below ⤵️

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

Project Thread: pbNhbs-bJN-p2

Design Thread: p1HpG7-umC-p2

## Does this pull request change what data or activity we track or use?

No

## Testing instructions:

**To test this PR, skip down to "Testing this PR":**
Otherwise, if testing issue: _Backup Card: My Jetpack shows "Needs Plan" for backup even when there is a Jetpack plan from a hosting partner_, follow the instructions below:

#### Testing issue https://github.com/Automattic/jetpack/issues/39762 :<br/>(Backup Card: My Jetpack shows "Needs Plan" for backup even when there is a Jetpack plan from a hosting partner.)
- Create a Pressable hosting partner site (if you don't already have one) by following instructions here: PCYsg-9TF-p2
    - (sign up using the direct link to the a8c employee plan that requires no coupon)
- Install the Jetpack Beta Tester plugin: https://github.com/Automattic/jetpack-beta
- Activate this PR branch for Jetpack in Jetpack Beta plugin.
- Purchase a Backup plan.
- Open My Jetpack and confirm that Backup product card status is Active and not "Needs plan". 

#### Testing this PR:
* Test this PR with a Jurassic Ninja Site running the Beta plugin on the `add/mj-expired-products-cards` branch. In the JN interface, you can automatically assign paid product licenses for all the products shown on the My Jetpack page, See screenshot:

Screen Shot on 2024-10-22 at 22-08-12
* Connect Jetpack, both site and user.  (Once user connected, you will see all the paid plans in My Jetpack.)
* Confirm that all paid products are showing the correct status (For example confirm Backup and Search are not showing "Needs plan" when the paid plan has actually been purchased) - See issue: https://github.com/Automattic/jetpack/issues/39762
* Now look up your site in Store Admin. 
* One by one, for each of the subscriptions, change the expiration date to either expiring (nearing expiration), or expired . You may need to disable 'auto-renew' for the plan to show as expiring/expired. The go to My Jetpack, reload the page and look at the product card for that plan.
* Verify the status is either "Expired" or "Expiring soon", depending on the date you set.
* Verify there is a red border around expired products, and a yellow border around products expiring soon.
* Verify that for products that are Expired, the CTA says "Resume my plan" and points to the wordpress.com purchase management page. For products that are Expiring soon, verify the CTA says "Renew my plan", and points directly to checkout with the product renewal in the cart.
* Continue this (set the expiration date, then view in My Jetpack) for each of the products.
* Once done testing all products, in Store Admin you can remove all the product subscriptions.
* Now Purchase a bundle for the site, such as Complete, or Security.
* In Store Admin, set the bundle subscription date to expired/expiring. (you may need to disable auto-renew).
* Now view the products included with the bundle on My Jetpack.  The products included with the bundle should be showing as expired/expiring soon, in My Jetpack.
* Verify that the expired/expiring CTA points to the bundle subscription in the purchase management page on wordpress.com (for expired), or points directly to checkout (for expiring-soon).
* Establish a site-only connection (no user-connection) and verify everything works properly with no errors. Some product cars may now show the status "Requires user connection", which takes precedence over expired/expiring products.

### CODE REVIEW

File: projects/packages/my-jetpack/_inc/components/product-card/style.module.scss
Comment (IanRamosC) [inline projects/packages/my-jetpack/_inc/components/product-card/style.module.scss]: I saw that the styles show that the borders are 2px wide, but I personally think that 1px looks better and less aggressive, even though 2px draws more attention to the cards:

2px | 1px
--- | --- 
image|image
image|image

File: projects/packages/my-jetpack/src/products/class-anti-spam.php
Comment (IanRamosC) [inline projects/packages/my-jetpack/src/products/class-anti-spam.php]: Would we like to include WPCOM plans that include Jetpack features or other old Jetpack plans (like Starter) here as well?

Comment (IanRamosC) [reply projects/packages/my-jetpack/src/products/class-anti-spam.php]: Considering that WPCOM plans would have a different checkout, etc

**Review by IanRamosC — DISMISSED** [review]
This looks pretty good. The changes are easy to follow and the code tested well. 👍🏾

**Comment by IanRamosC** [comment]
It seems like you need to update the PR before merging

File: projects/packages/my-jetpack/_inc/components/product-card/style.module.scss
Comment (elliottprogrammer) [reply projects/packages/my-jetpack/_inc/components/product-card/style.module.scss]: Ok yeah, sounds good. Thanks Ian, I changed the error & warning border to 1px. 👍

File: projects/packages/my-jetpack/src/products/class-anti-spam.php
Comment (elliottprogrammer) [reply projects/packages/my-jetpack/src/products/class-anti-spam.php]: Yeah, this comment/feedback is no longer relevant. I refactored this so now we're pulling the bundle slugs from the WPCOM features endpoint. I'm marking this comment as "Resolved". 👍

File: projects/packages/my-jetpack/src/products/class-product.php
Comment (CodeyGuyDylan) [inline projects/packages/my-jetpack/src/products/class-product.php:414]: Seems like having this check fulfills most of what [this issue](https://github.com/Automattic/jetpack-marketing/issues/905) needs to be fixed 😄  Once this is done, it should be a quick patch to get this one done finally 🥳

File: projects/packages/my-jetpack/src/products/class-creator.php
Comment (CodeyGuyDylan) [inline projects/packages/my-jetpack/src/products/class-creator.php:319]: The creator class won't really be doing anything now that Growth is here, so we probably don't need to update it. But I guess it's good for complete-ness

File: projects/packages/my-jetpack/src/products/class-crm.php
Comment (CodeyGuyDylan) [inline projects/packages/my-jetpack/src/products/class-crm.php:181]: Do we only have this function in CRM since CRM exists outside of our system?

**Review by CodeyGuyDylan — CHANGES_REQUESTED** [review]
I mentioned this is the [other PR](https://github.com/Automattic/jetpack/pull/40115) but wanted to make sure it didn't get lost. I don't think we should be showing the "expired" status if it's been expired for a very long time. My gut reaction is we probably shouldn't show it if it's been expired for longer than a month

I noticed the "needs connection" error shows as a red error. I'm wondering if it should be yellow since the status is also yellow? Also, if the "Needs connection" error has a red or maybe yellow border, should the "Needs plan" status also have one?

image

Other than those thoughts, the plan recognition upgrade seems to work great! Thank you for your hard work on this

**Comment by elliottprogrammer** [comment]
First, thanks very much for the reviews @CodeyGuyDylan! 😃
Oh, my apologies Dylan, I meant to address/answer these points/feedback from your first review.  

> I mentioned this is the [other PR](https://github.com/Automattic/jetpack/pull/40115) but wanted to make sure it didn't get lost. I don't think we should be showing the "expired" status if it's been expired for a very long time. My gut reaction is we probably shouldn't show it if it's been expired for longer than a month

For not showing a product as "expired" after it's been expired for a certain amount of time, here's my thoughts:
If we don't show when the product is expired (after a certain time), then what status do we show it as then? Just "Active"?, or "Needs plan"? For products that require a paid plan to use them, if we show them as Active, it won't actually be active because it's expired. But we show the product as "Needs plan" instead, I don't think the user can re-purchase the same plan until they "Remove" the expired one first.  Something about showing a particular product card (like Backup, or whatever) like nothing is wrong with it, but then when the user clicks on the card to use it, it's not working properly because the plan subscription has been expired. It just seems like it might be opening a big can of worms..
When a product subscription expires, the user really should either 1.- Renew the subscription, Or 2.- Remove it (in the purchase management section). If they Remove the subscription, the expiration status and notice goes away. If they Renew the subscription, the status and notice goes away.  Thats why when a product is expired, the Notice CTA, I pointed to the purchase/subscription management section on wordpress.com where the user can either "Remove" the subscription or "Renew" it. However if the product is expiring soon, the Notice CTA points straight to checkout with the product renewal in the cart.
I noticed that Atomic sites also don't hide the plan expiration notice (after any amount of time) either.  In fact Atomic sites have the expiration notice displaying all over the place, on many different pages all throughout wp-admin (Every one of the settings pages (14 pages), the WP Dashboard page, the Hosting -> home page, the Jetpack - Akismet page, and others I think.), where I believe the thinking is that the user is expected to either renew the subscription, or remove it, if they want the notice to go away. (see screenshots: Here's only a few, out of many, many:)

Screen Shot 2024-11-25 at 20 16 56
Screen Shot 2024-11-25 at 20 18 29
Screen Shot 2024-11-25 at 20 19 50

Anyway, so let me know your thoughts. Let me know if maybe I'm missing something, or maybe over-thinking it?

I did however make a change in the code where I added in the ability to hide/don't show the expiration status after 2 months after the expiration date. I didn't turn it on by default. But it would only be a matter of passing in a `true` value into a function call, if we should decide to go that route. 😉🤷‍♂️

> I noticed the "needs connection" error shows as a red error. I'm wondering if it should be yellow since the status is also yellow? Also, if the "Needs connection" error has a red or maybe yellow border, should the "Needs plan" status also have one?

Hmm, I changed/fixed this prior to your review.. I wonder if your branch (or Jetpack Beta branch) was not up to the latest?
I changed the product card border back to how it was originally, where connection warnings and other warnings/errors do not show a red or yellow border around the product cards. Only the expired/expiring products show the red/yellow borders around them. 🙂

File: projects/packages/my-jetpack/src/products/class-creator.php
Comment (elliottprogrammer) [reply projects/packages/my-jetpack/src/products/class-creator.php:319]: Yeah, I just made them all the same & consistent with any changes I made. 👍

File: projects/packages/my-jetpack/src/products/class-crm.php
Comment (elliottprogrammer) [reply projects/packages/my-jetpack/src/products/class-crm.php:181]: Yes, all we can check for CRM is if the site as a Jetpack Complete plan (because CRM is included with Jetpack Complete). That's all we were doing previously, so that's all we are doing here too, so nothing has changed really.

**Comment by CodeyGuyDylan** [comment]
> Thats why when a product is expired, the Notice CTA, I pointed to the purchase/subscription management section on wordpress.com where the user can either "Remove" the subscription or "Renew" it.

This is a good point and maybe fixes the issue I had with it 🤔 I don't know that, as a user, I would find it clear that the way to make the notification go away is to remove the subscription altogether, though I guess I would probably try it eventually. (I hate notifications 😆). I think we are probably okay to keep it how it is then if that's the case. Could we add to tracks the ability to see how long a subscription has been expired when viewing a card with an expired plan? That way we could check in tracks later to see if users are leaving that notification there for long periods of time or not later on.

> In fact Atomic sites have the expiration notice displaying all over the place

Personally I don't think that's very good UX either 😅. I am not every user but personally that would bug the heck out of me

**Comment by elliottprogrammer** [comment]
@CodeyGuyDylan ,

> I don't know that, as a user, I would find it clear that the way to make the notification go away is to remove the subscription altogether, though I guess I would probably try it eventually. (I hate notifications 😆)

Yeah, so true, I think _we all_ hate notifications! 😆

Looking into it and thinking about this further, I'm noticing there's a way to show a close(X) icon on the notice!  I'll look now to see how much it would take to close the expired/expiring notices with some persistence!  This would be ideal, I think.  Do you agree? 😃

> Could we add to tracks the ability to see how long a subscription has been expired when viewing a card with an expired plan? That way we could check in tracks later to see if users are leaving that notification there for long periods of time or not later on.

This is a great idea!  Yes, I'll do this! 🙌

> > In fact Atomic sites have the expiration notice displaying all over the place
> 
> Personally I don't think that's very good UX either 😅. I am not every user but personally that would bug the heck out of me

Yeah, I agree.

**Review by CodeyGuyDylan — DISMISSED** [review]
More of a question instead of a requested change. In the scenario below, Boost, Social, and Protect all have expired plans. I see that some statuses override the expired plan status such as these. I think that is fine, as it wouldn't work anyway without the plugin being installed or active, so the statuses shown are still integral to getting the plugin to work. I just wanted to make sure it was expected behavior 😄 
image

One note, and this probably isn't feasible, but I think it would be great if we could redirect them back to My Jetpack after renewing or removing a product after getting to `/me/purchases` from My Jetpack. Maybe we could do that sometime in the future

Other than that I think all of my issues have been addressed. Thank you for all the changes! I'll put my conversation with payments about the renewal stuff here since that answers the question as to why we don't need to cut off renewals older than a month old 😄 

Slack: p1732730061873749-slack-C02BEP610

**Comment by elliottprogrammer** [comment]
> More of a question instead of a requested change. In the scenario below, Boost, Social, and Protect all have expired plans. I see that some statuses override the expired plan status such as these. I think that is fine, as it wouldn't work anyway without the plugin being installed or active, so the statuses shown are still integral to getting the plugin to work. I just wanted to make sure it was expected behavior 

@CodeyGuyDylan, Yes, that is expected behavior. 
There are some statuses that have a higher priority and they override the expiring plans notices.  Yeah all the statuses have a particular, sensible order in the `Product::get_status()` function. 👍

File: projects/packages/my-jetpack/src/products/class-social.php
Comment (manzoorwanijk) [inline projects/packages/my-jetpack/src/products/class-social.php:188]: Should this contain all the legacy products that are no longer used anymore?

Social currently has only those `*_v1_*` products.

Comment (elliottprogrammer) [reply projects/packages/my-jetpack/src/products/class-social.php:188]: Hey @manzoorwanijk! 👋  Thanks for the question! 😃
> Should this contain all the legacy products that are no longer used anymore?
> Social currently has only those *_v1_* products.

To answer your question: Yes, it should contain ALL possible Social product slugs (including legacy products, but not including Bundle plans)
Yes, ultimately `has_paid_plan_for_product()` uses this function in question (`get_paid_plan_product_slugs()`) as well as `get_paid_bundles_that_include_product()` to get All the paid plan product_slugs and bundle plan product_slugs that include Social, to ultimately determine if the site has a paid plan for the product (in this case Social).
Some sites/users might still be renewing a legacy Social product that they bought previously, so we need to know if the site/user has _ANY_ Social paid plan (including legacy plans, or even bundled plans that include Social within it). Basically, if the site/user has _Any_ paid plan that includes Social, we want to know about it. 😉

Comment (manzoorwanijk) [reply projects/packages/my-jetpack/src/products/class-social.php:188]: Hey

Oh yeah. Thanks.

File: projects/packages/my-jetpack/src/products/class-product.php
Comment (robertsreberski) [reply projects/packages/my-jetpack/src/products/class-product.php:414]: @grzegorz-cp would you be able to follow up on this, and get [the issue](https://github.com/Automattic/jetpack-marketing/issues/905) done?

**Pull request merged**
